### ISSUE TITLE
document.documentElement and document.head are undefined on document-start

### ISSUE THREAD

**Issue by dumpsters** [greasemonkey/greasemonkey#2996]
Having a problem where sometimes document.head or even document.documentElement will be undefined when the script runs, doesn't happen 100% of the time.
![undefined](https://user-images.githubusercontent.com/12140220/46036535-3bf17300-c0f5-11e8-9845-d48f8a3cd6f6.png)
The following code is as close as possible to my real one after minifying, only difference is that asyncOperation
 is a httpRequest in mine so I made the timeout to simulate the time I measured on average.
I've been able to reproduce it by entering a website and (middle-mouse) clicking a bunch of links in said website then checking the tabs afterwards and their console.
```
// ==UserScript==
// @name                    TestRunAt
// @namespace               testrunat
// @run-at                  document-start
// @author                  LLLLLLLLLLLLLLLLLL
// @description             TEST
// @version                 1
// @include                 *
// ==/UserScript==


"use strict";

const mainHTML = document.documentElement || document.getElementsByTagName("html")[0];

mainHTML && (mainHTML.style.display = "none");

class TestClass {
    constructor() {}
    addStyle(e) {
        let t = document.head || document.getElementsByTagName("head")[0];
        if (!t) return console.error("Head: %s", typeof t), console.error("HTML: %s", typeof mainHTML), 
        null;
        {
            let n = document.createElement("style");
            n.setAttribute("type", "text/css"), n.textContent = e, t.appendChild(n), mainHTML.style.display = "block";
        }
    }
    asyncOperation(e) {
        let timeSimulate = Math.floor(Math.random() * 300) + 100)
        return new Promise(t => {
            setTimeout(() => t(e), (timeSimulate);
        });
    }
}

TestClass = new TestClass();

TestClass.asyncOperation("*{background: #404040 !important;}").then(e => TestClass.addStyle(e)).catch(e => console.error(e));

document.addEventListener("DOMContentLoaded", () => {
    // Do DOM related stuff
}, !1);
```

**Comment by arantius** [comment]
https://wiki.greasespot.net/Greasemonkey_Manual:Getting_Help#What_to_Say

Thank you for providing most of this, especially the reduced test case.  But: Firefox, Greasemonkey version are missing and feel important in this case.

**Comment by dumpsters** [comment]
Hey sorry about that I've edited the original post to include the greasemonkey version and more firefox details

**Comment by Sxderp** [comment]
I don't think this is really a bug. When you run your script at 'document-start' (disregarding the message passing that GM does which artificially delays injection) it gets injected really early. I haven't found specifics for Firefox, but for Chrome this occurs before the DOM is created. Firefox is likely similar and from what I've seen there's no guarantee on what (if any) part of the DOM will be available during execution.

See https://code-examples.net/en/q/293af5b which discusses some of Chrome's behaviors.

If Greasemonkey ever moves to the `contentScript`  or `userScript` API then this 'issue' will just become more prevalent because those APIs would remove the need for GM to pass messages in order to run the scripts, thus removing the artificial delay.

As for solving your issue; you're going to have to delay your actions with the DOM (style injection) until the proper elements are loaded. You can use an event listener on a standard event, or a promise loop that checks the element on each iteration.

**Comment by arantius** [comment]
Back in 3.x we supported `document-start` and specifically engineered it to make sure there was some sort of DOM available (for e.g. CSS injection).  In 4.x we don't even support `document-start` so *shrug* this is hard to make sense of.

**Comment by Sxderp** [comment]
Current Greasemonkey doesn't guarantee `document-start` but that doesn't mean scripts won't execute at (or come close to)`document-start`.  It's more of a "execute this script as soon as possible" wherein the other execution times, by their nature of being later in the pipeline, are closer to "don't execute this script until 'x' time". 

There's probably some combination of FF code changes and this user's environment that happens to cause GM to request / execute the script(s) some number of ticks earlier than the average (as measured some time ago). Resulting in behavior that more closely matches the 'proper' (as defined by MDN and Chrome documentation) `document-start` environment. 

As to what to do. At this moment I can't think of anything promising. I don't know how 3.x did whatever but it's likely that's not available any longer. And whatever solution is proposed would still need to take into account the indeterministic state of `document-start`.
